File: src/main/kotlin/org/gameflow/entity/EntityGroup.kt
package org.gameflow.entity

import java.util.ArrayList
import org.gameflow.util.ListenableCollection
import java.util.Collection
import org.gameflow.RenderContext
import java.util.concurrent.ConcurrentSkipListSet
import java.util.Collections
import org.gameflow.component.Component
import org.gameflow.pass.Passable
import org.gameflow.pass.Pass
import org.gameflow.Game

/**
 * A group of entities.
 */
public open class EntityGroup(name: String): Passable {

    private val entities = ConcurrentSkipListSet<Entity>()
    private val unmodifiableEntities: Collection<out Entity> = Collections.unmodifiableCollection<Entity>(entities) !!

    public final fun entities() : Collection<out Entity> {
        return unmodifiableEntities
    }

    public final fun add(entity: Entity)  {
        entities.add(entity)
        onEntityAdded(entity)
    }

    public final fun remove(entity: Entity)  {
        entities.remove(entity)
        onEntityRemoved(entity)
    }

    public final fun clear() {
        onAllEntitiesRemoved(unmodifiableEntities)
        entities.clear()
    }

    protected open fun onEntityAdded(entity: Entity) {}
    protected open fun onEntityRemoved(entity: Entity) {}
    protected open fun onAllEntitiesRemoved(entities: Collection<out Entity>) {}


    override fun doPassOnChildren(pass : Pass, game : Game) {
        for (entity in entities) {
            entity.doPass(pass, game)
        }
    }
}

File: src/main/scala/org/gameflow/core/entity/EntityGroup.scala
package org.gameflow.core.entity

import java.util.concurrent.ConcurrentSkipListSet
import java.util.Collections
import org.gameflow.core.pass.Passable
import org.gameflow.core.pass.Pass
import org.gameflow.core.Game
import scala.collection.JavaConversions._

/**
 *
 */
class EntityGroup(val name: Symbol) extends Passable {

  private val _entities = new ConcurrentSkipListSet[Entity]()
  private val unmodifiableEntities = Collections.unmodifiableCollection[Entity](_entities)

  final def entities : java.util.Collection[Entity] = {
    unmodifiableEntities
  }

  final def add(entity: Entity) {
    _entities.add(entity)
    onEntityAdded(entity)
  }

  final def remove(entity: Entity) {
    _entities.remove(entity)
    onEntityRemoved(entity)
  }

  final def clear() {
    onAllEntitiesRemoved(unmodifiableEntities)
    _entities.clear()
  }

  protected def onEntityAdded(entity: Entity) {}
  protected def onEntityRemoved(entity: Entity) {}
  protected def onAllEntitiesRemoved(entities: java.util.Collection[Entity]) {}

  override final def doPassOnChildren(pass : Pass, game : Game) {
    _entities.foreach {_.doPass(pass, game)}
  }

}